test/linalg/qr.jl

```diff
@@ -38,107 +37,108 @@ for eltya in (Float32, Float64, Complex64, Complex128, BigFloat, Int)
-                @test q*r*qrpa[:P].' ≈ a[1:n1,:]
-                @test_approx_eq_eps a[1:n1,:]*(qrpa\b[1:n1]) b[1:n1] 5000ε
-                @test full(qrpa) ≈ a[1:5,:]
-                @test_throws DimensionMismatch q*b[1:n1+1]
-                @test_throws DimensionMismatch b[1:n1+1]*q'
-                if eltya != Int
-                    @test eye(eltyb,n1)*q ≈ convert(AbstractMatrix{tab},q)
+                @testset "(Automatic) Fat (pivoted) QR decomposition" begin
+                    @inferred qrfact(a, Val{true})
+                    @inferred qr(a, Val{true})
+
+                    qrpa  = factorize(a[1:n1,:])
+                    q,r = qrpa[:Q], qrpa[:R]
+                    @test_throws KeyError qrpa[:Z]
+                    p = qrpa[:p]
+                    @test q'*full(q, thin=false) ≈ eye(n1)
+                    @test q*full(q, thin=false)' ≈ eye(n1)
+                    @test (UpperTriangular(eye(eltya,size(q,2)))*q')*full(q, thin=false) ≈ eye(n1)
+                    @test q*r ≈ (isa(qrpa,QRPivoted) ? a[1:n1,p] : a[1:n1,:])
+                    @test q*r[:,invperm(p)] ≈ a[1:n1,:]
+                    @test q*r*qrpa[:P].' ≈ a[1:n1,:]
+                    @test_approx_eq_eps a[1:n1,:]*(qrpa\b[1:n1]) b[1:n1] 5000ε
+                    @test Array(qrpa) ≈ a[1:5,:]
+                    @test_throws DimensionMismatch q*b[1:n1+1]
+                    @test_throws DimensionMismatch b[1:n1+1]*q'
+                    if eltya != Int
+                        @test eye(eltyb,n1)*q ≈ convert(AbstractMatrix{tab},q)
+                    end
                 end
-
-debug && println("(Automatic) Thin (pivoted) QR decomposition")
-                qrpa  = factorize(a[:,1:n1])
-                q,r = qrpa[:Q], qrpa[:R]
-                @test_throws KeyError qrpa[:Z]
-                p = qrpa[:p]
-                @test q'*full(q, thin=false) ≈ eye(n)
-                @test q*full(q, thin=false)' ≈ eye(n)
-                @test q*r ≈ a[:,p]
-                @test q*r[:,invperm(p)] ≈ a[:,1:n1]
-                @test full(qrpa) ≈ a[:,1:5]
-                @test_throws DimensionMismatch q*b[1:n1+1]
-                @test_throws DimensionMismatch b[1:n1+1]*q'
-                @test_approx_eq_eps A_mul_Bc(UpperTriangular(eye(eltyb,size(q.factors,2))),q)*full(q, thin=false) eye(n1,n) 5000ε
-                if eltya != Int
-                    @test eye(eltyb,n)*q ≈ convert(AbstractMatrix{tab},q)
+                @testset "(Automatic) Thin (pivoted) QR decomposition" begin
+                    qrpa  = factorize(a[:,1:n1])
+                    q,r = qrpa[:Q], qrpa[:R]
+                    @test_throws KeyError qrpa[:Z]
+                    p = qrpa[:p]
+                    @test q'*full(q, thin=false) ≈ eye(n)
+                    @test q*full(q, thin=false)' ≈ eye(n)
+                    @test q*r ≈ a[:,p]
+                    @test q*r[:,invperm(p)] ≈ a[:,1:n1]
+                    @test Array(qrpa) ≈ a[:,1:5]
```

Comment:
Re. `Array(qrpa)`, please see the comment above regarding `full` translating to `AbstractArray` for factorizations.